### ISSUE TITLE
HeidiSQL 12.8.0.6908 contains vulnerable OpenSSL DLLs

### ISSUE THREAD

**Issue by AScott-WWF** [HeidiSQL/HeidiSQL#2022]
### What where you trying to do?
Identifying applications in use in our organisation containing vulnerable 3rd party components which affect our organisations vulnerability risk / score

### What actually happened?
I have identified that the latest HeidiSQL installer v12.8.0.6908 contains OpenSSL v1.1.1s and v3.1.5 DLLs
OpenSSL v1.1.1s is vulnerable to at least 16 CVEs (The highest severity is HIGH).
OpenSSL v3.1.5 is vulnerable to at least 5 CVEs (1 MODERATE severity & 4 LOW severity).

With HeidiSQL installed for all users (in C:\Program Files\HeidiSQL), using the following PowerShell, here is the evidence:
```
cd 'C:\Program Files\HeidiSQL\'
Get-ChildItem *libcrypt*.dll,*libssl*.dll,*openssl.exe,heidisql.exe -Recurse -Force -ErrorAction SilentlyContinue | Select-Object * -ExpandProperty versioninfo | Sort-Object ProductVersion,FileVersionRaw,Filename | Select-Object ProductVersion,FileVersionRaw,Filename,FileDescription,CompanyName,LegalCopyright | ft -auto
```
**Results:**
```
ProductVersion FileVersionRaw FileName                                        FileDescription CompanyName                                   LegalCopyright
-------------- -------------- --------                                        --------------- -----------                                   --------------
               12.8.0.6908    C:\Program Files\HeidiSQL\heidisql.exe
1.1.1s         1.1.1.19       C:\Program Files\HeidiSQL\libcrypto-1_1-x64.dll OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2022 The OpenSSL Authors. All rights reserved.
1.1.1s         1.1.1.19       C:\Program Files\HeidiSQL\libssl-1_1-x64.dll    OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2022 The OpenSSL Authors. All rights reserved.
3.1.5          3.1.5.0        C:\Program Files\HeidiSQL\libcrypto-3-x64.dll   OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2024 The OpenSSL Authors. All rights reserved.
3.1.5          3.1.5.0        C:\Program Files\HeidiSQL\libssl-3-x64.dll      OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2024 The OpenSSL Authors. All rights reserved.
```

As OpenSSL release new versions from time to time to fix published CVEs, it is important that software vendors keep any 3rd party components (OpenSSL DLLs in this case) up to date to prevent these vulnerabilities giving malicious code an access point.

The most recent OpenSSL release was on the 3rd September (v3.0.15, v3.1.7, v3.2.3 and v3.3.2) to fix some recently published CVEs
Full List of OpenSSL Vulnerabilities: https://openssl-library.org/news/vulnerabilities/
OpenSSL v1.1.1 is now end of life so should no longer be used (although security updates are available for customers with a [premium level support agreement](https://openssl-corporation.org/support/))

### What did you expect to happen?
Please remove OpenSSL v1.1.1 from the HeidiSQL product and update to v3.1.7 (or either v3.2.3 or v3.3.2)

### HeidiSQL version

12.8.0.6908

### Database server version

-

### Error/Backtrace

```shell
n/a
```

### Reproduction recipe

PowerShell evidence:
```
cd 'C:\Program Files\HeidiSQL\'
Get-ChildItem *libcrypt*.dll,*libssl*.dll,*openssl.exe,heidisql.exe -Recurse -Force -ErrorAction SilentlyContinue | Select-Object * -ExpandProperty versioninfo | Sort-Object ProductVersion,FileVersionRaw,Filename | Select-Object ProductVersion,FileVersionRaw,Filename,FileDescription,CompanyName,LegalCopyright | ft -auto
```
**Results:**
```
ProductVersion FileVersionRaw FileName                                        FileDescription CompanyName                                   LegalCopyright
-------------- -------------- --------                                        --------------- -----------                                   --------------
               12.8.0.6908    C:\Program Files\HeidiSQL\heidisql.exe
1.1.1s         1.1.1.19       C:\Program Files\HeidiSQL\libcrypto-1_1-x64.dll OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2022 The OpenSSL Authors. All rights reserved.
1.1.1s         1.1.1.19       C:\Program Files\HeidiSQL\libssl-1_1-x64.dll    OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2022 The OpenSSL Authors. All rights reserved.
3.1.5          3.1.5.0        C:\Program Files\HeidiSQL\libcrypto-3-x64.dll   OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2024 The OpenSSL Authors. All rights reserved.
3.1.5          3.1.5.0        C:\Program Files\HeidiSQL\libssl-3-x64.dll      OpenSSL library The OpenSSL Project, https://www.openssl.org/ Copyright 1998-2024 The OpenSSL Authors. All rights reserved.
```

**Comment by ansgarbecker** [comment]
32bit version is affected as well I suppose. Some of these files, for instance libcrypto-1_1-32.dll is a dependency for libpq.dll v10.x, which I would then have to drop as well. libpq v10 is the last precompiled 32bit version available out there. Means HeidiSQL will come without 32bit support for PostgreSQL as soon as I release. Hm..

**Comment by AScott-WWF** [comment]
Hmm, PostgreSQL v10 is the only version to support 32bit Windows (2008 R1, 7, 8, 10) https://www.postgresql.org/download/windows/)
TBH I'd be surprised if there are many PostgreSQL users would still be using a 32-bit OS in 2024!
Maybe it's time to modernise 😂

**Comment by AScott-WWF** [comment]
Yep, entirely agree, many in the 3rd world doesn't have the luxury of always having the latest modern hardware.
Maybe a potential solution is keep the OpenSSL version current (3.1.7 currently) for 64 bit systems and plan for the removal of 32bit OpenSSL (v1.1.1) to coincide with the end of support for Windows 10 (in October 2025)?
Or split out the installers to 64bit and 32 bit, so 64bit versions drop 32 bit support (and therefore do not contain the vulnerable OpenSSL v1.1.1 DLLs)?
We personally only use the 64 bit version of HeidiSQL and would prefer to have the latest supported OpenSSL DLLs to prevent a known security vulnerability causing us issues.
But I fully accept, it's your great product and your decision how to solve.

**Comment by ansgarbecker** [comment]
I asked the users in May 2021 whether someone still needs 32bit support: https://www.heidisql.com/forum.php?t=37945 . I had a look at the statistics at that point, and found roughly 1% of all users were on 32bit.
Now, 3 years later, we have roughly 0.5%, so even less.

I should probably drop 32bit support from the releases, and just keep it for portables. On the other side, 32bit users could still use an old version which I am still offering for download, so I could anyway drop them entirely.

**Comment by AScott-WWF** [comment]
That sounds a very pragmatic approach.
Looking forward to a fixed version in due course

Thanks @ansgarbecker

**Comment by ansgarbecker** [comment]
Done. The upcoming installers will have the current OpenSSL libraries in both 1.x and 3.x branches. Additionally, I updated libpq.dll v15, and added libpq.dll v17.
